### ISSUE TITLE
Can't read large files without massive memory consumption

### ISSUE THREAD

**Issue by McCrafterIV** [brendan-duncan/archive#182]
When you want to extract or open large zip files (couldn't test it with other formats, but I think this applies to all supported archives) the whole file is loaded into memory, which can quickly be a problem if you deal with large files (2GB or above) since the decoder doesn't accept any kind of streams. It would be great if you at least could extract an archive without massive memory usage. Is there a way to solve this issue or any plans to implement this in the foreseeable future?

**Comment by brendan-duncan** [comment]
There is a InputFileStream class that can be passed to decodeBuffer, to at least alleviate the memory for storing the compressed buffer. I had started working on file->file streaming, but I don't think I ever finished that. I should probably look into that.

**Comment by McCrafterIV** [comment]
Could you provide a quick example on how to use InputFileStream?
File to file streaming would be great, but of you find a way to open and work with an archive without reading the whole file from disk, this probably wouldn't even be needed anymore (like 7zip and others can opem archives without loading the whole file into memory).

**Comment by brendan-duncan** [comment]
You can look at the io test, https://github.com/brendan-duncan/archive/blob/master/test/tests/io_test.dart#L139
`test('stream tar decode', () { ...`
It should be the same use with `ZipDecoder`. You can import those classes with `import 'package:archive/archive_io.dart';`

At one point I was looking at improving the support for io and streaming. Work and life then took over most of my time, and I never got back to cleaning it up or adding missing features.

**Comment by McCrafterIV** [comment]
Thanks for your quick response! I sadly can't get it to work with `InputFileStream` and `decodeBuffer`. I created a simple tar file, which works fine for both, `decodeBuffer` and `decodeBytes`. But when trying to load a simple zip file (same content as in the tar file), only decodeBytes works, decodeBuffer fails with `FormatException: Could not find End of Central Directory Record
` To test, if this is an issue regarding my specific zip file, I also used [readme.zip](https://github.com/brendan-duncan/archive/blob/master/test/tests/res/zip/readme.zip) from the tests, which resulted in the same exception.
```
FormatException: Could not find End of Central Directory Record
#0      ZipDirectory._findSignature (package:archive/src/zip/zip_directory.dart:149:5)
#1      new ZipDirectory.read (package:archive/src/zip/zip_directory.dart:28:20)
#2      ZipDecoder.decodeBuffer (package:archive/src/zip_decoder.dart:19:30)
#3      FileHandler.open (package:project/file.dart:93:18)
<asynchronous suspension>
```
This is my code:
```dart
var stream = InputFileStream(path);
var archive = ZipDecoder();
archive.decodeBuffer(stream);
```

This issue was mentioned a few times before, but I couldn't find a solution in the other posts. Besides that, since I used the readme.zip which was directly downloaded from Github, the zip file shouldn't be the problem?

> At one point I was looking at improving the support for io and streaming. Work and life then took over most of my time, and I never got back to cleaning it up or adding missing features.

Maybe sometime, I think this package would really profit from it.

**Comment by brendan-duncan** [comment]
Looks like that InputFileStream class I wrote is garbage for random access, which zip needs :-)
I'll get it fixed up.

**Comment by brendan-duncan** [comment]
I pushed some fixes to git for getting InputFileStream to work with ZipDecoder.

I also added an ArchiveFile.clear method that will clear the decompressed memory of an archive file. So, while not perfect file->file streaming, you can at least reduce memory a little bit more now.
```dart
// InputFileStream only uses a cache buffer memory (4k by default), not the entire file
var stream = InputFileStream('test.zip');
// The archive will have the memory of the compressed archive. ArchiveFile's are decompressed on
// demand
var zip = ZipDecoder().decodeBuffer(stream);
for (var file in zip.files) {
  final f = File(file.name);
  f.createSync(recursive: true);
  // file.content will trigger a decompression of the file, allocating it's memory
  f.writeAsBytesSync(file.content);
  // file.clear() will release the file's compressed memory
  file.clear();
}
```

**Comment by McCrafterIV** [comment]
Thanks for the work! I tested the new InputFileStream and it works great with big zip files. The memory usage is at least in my short tests negligible. It would be great if you could also create a release on pub.dev so no one needs to add the git repo to their pubspec.yaml :-)

**Comment by brendan-duncan** [comment]
Thanks for confirming. I'll get it released tonight (PST). I want to investigate having ArchiveFile use InputFileStream for it's compressed buffer references, which would keep all compressed data from having to be loaded into memory. But I'll get this released first.

**Comment by brendan-duncan** [comment]
Published, 3.1.9

**Comment by alesimula** [comment]
@brendan-duncan Hello, I have other few observations about performance and memory usage:

I tried these two different approaches for opening the test file I linked above:

```
var stream = InputFileStream(file.absolute.path);
return ZipDecoder().decodeBuffer(stream);
```
and
```
return ZipDecoder().decodeBytes(file.readAsBytesSync());
```

And noticed the second one, where I just passed the UInt8List obtained from readAsBytesSync, seems way faster than using the InputFileStream.

Initial memory usage also seemed to be way lower with the second approach, though, after some time, all memory used by the first method got cleaned up by the garbage collector, while the memory occupied using the second method would stay open indefinitely until I closed the application (hence why I also opened [this issue](https://github.com/brendan-duncan/archive/issues/180)).

Provided I don't know how the library (or compression) works, would it be possible to initially just read the archive header instead of its whole content, and make it so it would read the files "on demand" (so it would just store the initial and final offsets of each file's content)?
